### DISCUSSION TITLE
How do I run without NPM?

### DISCUSSION THREAD

**Discussion by festing** [t1m0n/air-datepicker#621] [General]
on shared server, no access to install via NPM.  Downloaded src files.  How do I use?

**Comment by il97mur** [comment]
Hi @festing , perhaps you can try using the files from CDN:
https://www.jsdelivr.com/package/npm/air-datepicker

**Comment by festing** [comment]
Great.  that works... NOW how do I change language to english?  Documentation method does not work as it seems it's refering to NPM install object.   I can get other options to work but NOTHING I try changes to english.

**Comment by il97mur** [comment]
@festing You can create an locale object before datepicker initialization and use it.
For example:
```
<script>
        const input = document.querySelector('.datepicker');

        const localeObj = {
            days: ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'],
            daysShort: ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'],
            daysMin: ['Su', 'Mo', 'Tu', 'We', 'Th', 'Fr', 'Sa'],
            months: ['January','February','March','April','May','June', 'July','August','September','October','November','December'],
            monthsShort: ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'],
            today: 'Today',
            clear: 'Clear',
            dateFormat: 'mm/dd/yyyy',
            timeFormat: 'hh:ii aa',
            firstDay: 0
        };

        const datepicker = new AirDatepicker(input, {
            locale: localeObj,
        });
</script>
```
 
The locale object is a copy from here: https://github.com/t1m0n/air-datepicker/blob/v3/dist/locale/en.js

**Reply by trymeouteh** [reply]
Thank you for this.

I do hope there is a UMD build in the future of Air Datepicker to not need to manually create the locale translations.

**Comment by festing** [comment]
Hmmm... seems like there should be a way to somehow link to the existing language JS in script, rather than having to open and then copy/paste.

**Comment by Pathologic** [comment]
@festing 
```
<!doctype html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport"
          content="width=device-width, user-scalable=no, initial-scale=1.0, maximum-scale=1.0, minimum-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Document</title>
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/t1m0n/air-datepicker@v3/dist/air-datepicker.css">
</head>
<body>

<input id="air_datepicker" type="text">

<script>
    var exports = {
      default: {}
    }
</script>

<script src="https://cdn.jsdelivr.net/gh/t1m0n/air-datepicker@v3/dist/air-datepicker.js"></script>
<script src="https://cdn.jsdelivr.net/gh/t1m0n/air-datepicker@v3/dist/locale/bg.js"></script>
<script>
  new exports.AirDatepicker('#air_datepicker', {
    locale: exports.default
  })
</script>
</body>
</html>
```

**Reply by lewysigns** [reply]
Thank you Pathologic!

Could you please explain why the following portion is necessary? (I'd just like to understand before I implement)

```
<script>
    var exports = {
      default: {}
    }
</script>
```

**Reply by lewysigns** [reply]
Had a discussion in gpt to answer my own question: 

"The exports object is initially defined in the HTML file. When the external scripts are loaded and executed, they attach the AirDatepicker function and locale settings to the exports object, which allows the final script to initialize the datepicker with the correct settings."

Looking at the code it looks like this is correct. Thank you @Pathologic , did you learn this via documentation or reading through the code? (Only ask because I'd like to read the documentation if it exist)

**Reply by MagnusOxlund** [reply]
@lewysigns I've brought the topic up again in the existing issue [#440](https://github.com/t1m0n/air-datepicker/issues/440#issuecomment-2365054115).

**Comment by jpgateaux** [comment]
Rewrote local files and placed them in locales directory
[locales.zip](https://github.com/user-attachments/files/18923868/locales.zip)


<h3>in en.js</h3>

<pre>
"use strict";


locales.en = void 0;
var _default = {
  days: ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'],
  daysShort: ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'],
  daysMin: ['Su', 'Mo', 'Tu', 'We', 'Th', 'Fr', 'Sa'],
  months: ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'],
  monthsShort: ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'],
  today: 'Today',
  clear: 'Clear',
  dateFormat: 'MM/dd/yyyy',
  timeFormat: 'hh:mm aa',
  firstDay: 0
};
locales.en = _default;
</pre>
<h3>in html</h3>
<pre>
&lt;script&gt;
    var locales = {};
&lt;/script&gt;

&lt;script src="./air-datepicker/locales/en.js">&lt;/script&gt;//one for each locale on page
</pre>
<h3>in js</h3>
<pre>
new AirDatepicker('#air_datepicker', {
    locale: locales.en
  })
</pre>

this way you can have different locales on same page without copy and paste

**Comment by racilhilan** [comment]
@festing, if you use a local copy as in your original question, then the best is to follow the recommendation from @il97mur or the enhanced version from @jpgateaux, although I wouldn't use the intermediary "_default" variable, simply assign the value directly to the "locales" object: `locales.en = {...}`.

To use a CDN, you'll need a workaround similar to the one from @Pathologic. However, note the following:

- The main object "AirDatepicker" doesn't need to attach itself to an object called "exports" as ChatGPT explained to you. That's incorrect. The reason it does that when you provide an "exports" object like @Pathologic did is that it passes the check `"object" == typeof exports` near the top of the file and so it branches to `exports.AirDatepicker = t()`. This test is actually meant for a module, but providing an object makes it pass the check. If you don't provide the object, the check will fail and correctly branches to `e.AirDatepicker = t())` which attaches it globally to `window` (in this context, `e` equals `this` which is `window`). So it is recommended not to provide the "exports" object before the main script. Here is how the code above looks like altogether:
```
"object" == typeof exports
        ? (exports.AirDatepicker = t())
        : (e.AirDatepicker = t());
```
- The locale settings will try to attach themselves to an object called "exports" which doesn't exists because you're loading the files with `<script>` tags. The script is meant to be run as a module and actually export the locale (adding it to the exports). So as a workaround, you can provide the "exports" object (can be empty) before you load the locale script.
- The important part is that the main object "AirDatepicker" keeps a default locale in its "defaults" property, so you can access it using `AirDatepicker.defaults.locale`. It uses this value to populate the locale when you call it to enhance the element. You can set it to the value that was saved into the "exports.default" object or the intermediary "_default" object, then you can clear both objects to keep the code clean.

Here is an example with the workaround:

```
<!doctype html>
<html lang="en">
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/air-datepicker@3.6.0/air-datepicker.css">
</head>
<body>

<input id="calendar" type="text" />

<script src="https://cdn.jsdelivr.net/npm/air-datepicker@3.6.0/air-datepicker.js"></script>
<script type="text/javascript">
    var exports = {};
</script>
<script src="https://cdn.jsdelivr.net/npm/air-datepicker@3.6.0/locale/en.js"></script>
<script type="text/javascript">
    AirDatepicker.defaults.locale = _default;
    _default = undefined;
    exports = undefined;
</script>
<script type="text/javascript">
    new AirDatepicker('#calendar');
</script>
</body>
</html>
```

If you want to load multiple locales from the CDN, you can extend the idea from @jpgateaux and add them to the "locales" object one by one like this:

```
<script src="https://cdn.jsdelivr.net/npm/air-datepicker@3.6.0/air-datepicker.js"></script>
<script type="text/javascript">
    var exports = {};
    var locales = {};
</script>
<script src="https://cdn.jsdelivr.net/npm/air-datepicker@3.6.0/locale/en.js"></script>
<script type="text/javascript">
    locales.en= _default;
</script>
<script src="https://cdn.jsdelivr.net/npm/air-datepicker@3.6.0/locale/fr.js"></script>
<script type="text/javascript">
    locales.fr = _default;
    _default = undefined;
    exports = undefined;
</script>
```
Then you can pass the locale that you want when enhancing the element:

`new AirDatepicker('#air_datepicker', { locale: locales.en });`
